pkg/reconciliation/generic.go

```diff
@@ -18,8 +18,18 @@ type Reconcileable[T any] interface {
 	*T
 }
 
-// Try with U, a type of any whose POINTER still fulfils Reoncilable...
+// ReconcileObject ensures that desiredObject exists in the given state, either by creating it, or updating it if it
+// already exists.
 func ReconcileObject[U any, T Reconcileable[U]](ctx context.Context, kClient client.Client, requeueDelay time.Duration, desiredObject U) result.ReconcileResult {
+	recResult, _ := ReconcileAndGetObject[U, T](ctx, kClient, requeueDelay, desiredObject)
+	return recResult
+}
+
+// ReconcileAndGetObject ensures that desiredObject exists in the given state, either by creating it, or updating it if
+// it already exists. It returns the current state of the object on the server after the reconciliation.
+func ReconcileAndGetObject[U any, T Reconcileable[U]](
+	ctx context.Context, kClient client.Client, requeueDelay time.Duration, desiredObject U,
+) (result.ReconcileResult, *U) {
```

Comment:
I propose we add this variant, it's almost the same but returns the state of the object on the server after the reconciliation.
This is useful in two cases:
- the hash annotation matched so the object was not updated, but we're interested in a field that was updated independently by another controller. For example the IP of a LoadBalancer service in its status.
- we're interested in a field that is set by the write operation itself, for example `CreationTimestamp`.

Reply:
I think this solves some problems but might be overoptimising - it introduces more complexity than I'd like and relies on folks understanding the caching behaviour. Based on the conversation here I think that is tricky for even experienced users and I'd rather avoid callers having to make more decisions about those subtleties.

Reply:
> relies on folks understanding the caching behaviour

How so? We're either returning the result of a read that observed matching hash annotations, or the result of a create or update (not subject to caching).

If we don't have this variant, we'll need to resort to `ReconcileObject` followed by a separate `Get`. If anything, it sounds like it could produce more caching issues (Get not seeing the result of a write in ReconcileObject).

Reply:
> If we don't have this variant, we'll need to resort to ReconcileObject followed by a separate Get. If anything, it sounds like it could produce more caching issues (Get not seeing the result of a write in ReconcileObject).

In that behavior, we would need to reconcile and wait for the cache to have that object (and potentially reconcile again, fail etc). Otherwise, we need to use the Create/Update one, but never Create+Get or Update+Get combo.

So it makes sense to get the object back if user knows what to do with it. Actually, why would we even have any non-returning version? Create/Update always return the object also.